packages/astro/src/content/mutable-data-store.ts

```diff
@@ -6,9 +6,23 @@ import { AstroError, AstroErrorData } from '../core/errors/index.js';
 import { IMAGE_IMPORT_PREFIX } from './consts.js';
 import { type DataEntry, ImmutableDataStore } from './data-store.js';
 import { contentModuleToId } from './utils.js';
+import { tmpdir } from 'node:os';
+import { join } from 'node:path';
 
 const SAVE_DEBOUNCE_MS = 500;
 
+// Write a file atomically, without triggering the file watcher
+async function writeFileAtomic(filePath: PathLike, data: string) {
+	const dir = await fs.mkdtemp(tmpdir() + '/astro-');
+	const tempFile = join(dir, 'temp');
+	await fs.writeFile(tempFile, data);
+	try {
+		await fs.rename(tempFile, filePath);
+	} finally {
+		await fs.rmdir(dir).catch(() => {});
+	}
```

Comment:
I'm not too fond of this solution because now ALL projects, without discrimination, will incur in **four** (create temp dir, write file, rename file, remove directory) writing I/O operations instead of one. And writing operations are slow.

Maybe we could have an internal heuristic, where we enable this atomic rewriting only we exceed a certain number of collections.

Reply:
This is not a high-volume write though. It's debounced, so only called once in builds. Switching based on collection size is a good idea though.

Reply:
I've done a new approach which is just a single write and a rename. Renames are cheap, so it shouldn't be much overhead compared to the current approach.